swap/swap.go

```diff
@@ -434,86 +436,86 @@ func (s *Swap) Balances() (map[enode.ID]int64, error) {
 	return balances, nil
 }
 
-// SentCheque returns the last sent cheque for a given peer
-func (s *Swap) SentCheque(peer enode.ID) (cheque *Cheque, err error) {
-	if swapPeer := s.getPeer(peer); swapPeer != nil {
-		return swapPeer.getLastSentCheque(), nil
-	}
-	err = s.store.Get(sentChequeKey(peer), &cheque)
-	return cheque, err
-}
-
-// SentCheques returns the last sent cheques for all known SWAP peers
-func (s *Swap) SentCheques() (map[enode.ID]*Cheque, error) {
-	cheques := make(map[enode.ID]*Cheque)
+// Cheques returns all known last sent and received cheques, grouped by peer
+func (s *Swap) Cheques() (map[enode.ID]map[string]*Cheque, error) {
+	cheques := make(map[enode.ID]map[string]*Cheque)
 
+	// get peer cheques from memory
 	s.peersLock.Lock()
 	for peer, swapPeer := range s.peers {
 		swapPeer.lock.Lock()
-		cheques[peer] = swapPeer.getLastSentCheque()
+		cheques[peer] = make(map[string]*Cheque)
+		cheques[peer][sentChequeResponseKey] = swapPeer.getLastSentCheque()
+		cheques[peer][receivedChequeResponseKey] = swapPeer.getLastReceivedCheque()
 		swapPeer.lock.Unlock()
 	}
 	s.peersLock.Unlock()
 
-	// add store cheques, if peer was not already added
-	chequesIterFunction := func(key []byte, value []byte) (stop bool, err error) {
-		peer := keyToID(string(key), sentChequePrefix)
-		if _, peerHasCheque := cheques[peer]; !peerHasCheque {
-			var peerCheque Cheque
-			err = json.Unmarshal(value, &peerCheque)
-			if err == nil {
-				cheques[peer] = &peerCheque
-			}
-		}
-		return stop, err
+	// get peer cheques from store
+	err := s.addStoreCheques(sentChequePrefix, sentChequeResponseKey, cheques)
+	if err != nil {
+		return nil, err
 	}
-	err := s.store.Iterate(sentChequePrefix, chequesIterFunction)
+	err = s.addStoreCheques(receivedChequePrefix, receivedChequeResponseKey, cheques)
 	if err != nil {
 		return nil, err
 	}
 
-	return cheques, nil
-}
-
-// ReceivedCheque returns the last received cheque for a given peer
-func (s *Swap) ReceivedCheque(peer enode.ID) (cheque *Cheque, err error) {
-	if swapPeer := s.getPeer(peer); swapPeer != nil {
-		return swapPeer.getLastReceivedCheque(), nil
+	// fill in result with missing cheques
+	for _, peerCheques := range cheques {
```

Comment:
the result of the `Cheques` function is a map of maps. the keys at the outer level are peer IDs. then, for each peer, you have another map, which contains its last sent cheque and its last received cheque.

the purpose of this block of code is to always have at least a `nil` entry for each type of cheque, if the peer is present (i.e. has an entry in the map).

so the entry is not already empty, as there is no entry. peers could just have a partially empty map (which is not wrong, just another way of structuring the result). so this code aims to fill in these holes.

without it, the result looks something like this:
```
{
	"peerID1": {
		"sentCheque": <cheque>,
		"receivedCheque": <cheque>,
	},
	"peerID2": {
		"sentCheque": <cheque>,
	},
	"peerID3": {
		"receivedCheque": <cheque>,
	},
}
```
with it, the result looks like this:
```
{
	"peerID1": {
		"sentCheque": <cheque>,
		"receivedCheque": <cheque>,
	},
	"peerID2": {
		"sentCheque": <cheque>,
		"receivedCheque": null,
	},
	"peerID3": {
		"sentCheque": null,
		"receivedCheque": <cheque>,
	},
}
```
the idea here is to have a more structurally predictable result. 

but if you guys think it adds no or little value, i'll take it out. 

i am not so convinced myself, to be honest.

Reply:
of course. i was thinking of processing this result from outside (think js code that consumes this RPC, as @vojtechsimetka did for devcon).

anyway, i've removed this since i am less than fond of it at this point.